File: core/voice_capture/framing.py
"""HIKA binary framing codec (Python side of native helper protocol)."""

from __future__ import annotations

import struct
from typing import Optional, Tuple

from core.voice_capture.contracts import CaptureMessageType, DecodedCaptureFrame

MAGIC = b"HIKA"
VERSION = 1
HEADER_SIZE = 48
MAX_PAYLOAD = 65_536

_TYPE_MAP = {
    1: CaptureMessageType.READY,
    2: CaptureMessageType.PCM,
    3: CaptureMessageType.ERROR,
    4: CaptureMessageType.END,
    5: CaptureMessageType.CANCEL_ACK,
}
_TYPE_TO_INT = {v: k for k, v in _TYPE_MAP.items()}

# magic(4) ver(u16) type(u16) seq(u64) mono(u64) rate(u32) ch(u16) width(u16)
# plen(u32) reserved0(u32) reserved1(u32) reserved2(u32) => 48
_HEADER_STRUCT = struct.Struct("<4sHHQQIHHIIII")
assert _HEADER_STRUCT.size == HEADER_SIZE


def encode_frame(
    message_type: CaptureMessageType,
    *,
    sequence: int,
    monotonic_ns: int,
    payload: bytes = b"",
    sample_rate: int = 16_000,
    channels: int = 1,
    sample_width: int = 2,
) -> bytes:
    if not isinstance(message_type, CaptureMessageType):
        raise ValueError("invalid_message_type")
    if not isinstance(payload, (bytes, bytearray)):
        raise TypeError("payload_must_be_bytes")
    payload_b = bytes(payload)
    if len(payload_b) > MAX_PAYLOAD:
        raise ValueError("payload_too_large")
    if (
        isinstance(sequence, bool)
        or not isinstance(sequence, int)
        or sequence < 0
        or sequence > (2**64 - 1)
        or isinstance(monotonic_ns, bool)
        or not isinstance(monotonic_ns, int)
        or monotonic_ns < 0
        or monotonic_ns > (2**64 - 1)
    ):
        raise ValueError("invalid_numeric")
    if sample_rate != 16_000 or channels != 1 or sample_width != 2:
        raise ValueError("invalid_audio_format")
    if message_type == CaptureMessageType.PCM and (
        not payload_b or len(payload_b) % (channels * sample_width) != 0
    ):
        raise ValueError("invalid_pcm_payload")
    if message_type in {CaptureMessageType.END, CaptureMessageType.CANCEL_ACK} and payload_b:
        raise ValueError("unexpected_payload")
    if message_type == CaptureMessageType.ERROR and len(payload_b) > 64:
        raise ValueError("error_payload_too_large")
    if message_type == CaptureMessageType.READY and len(payload_b) > 1024:
        raise ValueError("ready_payload_too_large")
    type_int = _TYPE_TO_INT[message_type]
    header = _HEADER_STRUCT.pack(
        MAGIC,
        VERSION,
        type_int,
        sequence,
        monotonic_ns,
        sample_rate,
        channels,
        sample_width,
        len(payload_b),
        0,
        0,
        0,
    )
    assert len(header) == HEADER_SIZE
    return header + payload_b


def decode_header(data: bytes) -> Optional[Tuple[CaptureMessageType, int, int, int, int, int, int]]:
    if not isinstance(data, bytes) or len(data) < HEADER_SIZE:
        return None
    (
        magic,
        version,
        type_int,
        sequence,
        monotonic_ns,
        sample_rate,
        channels,
        sample_width,
        plen,
        r0,
        r1,
        r2,
    ) = _HEADER_STRUCT.unpack_from(data, 0)
    if magic != MAGIC or version != VERSION or r0 != 0 or r1 != 0 or r2 != 0:
        return None
    message_type = _TYPE_MAP.get(type_int)
    if message_type is None:
        return None
    if plen > MAX_PAYLOAD:
        return None
    if sample_rate != 16_000 or channels != 1 or sample_width != 2:
        return None
    return message_type, sequence, monotonic_ns, sample_rate, channels, sample_width, plen


def decode_frame(data: bytes) -> Optional[DecodedCaptureFrame]:
    parsed = decode_header(data)
    if parsed is None:
        return None
    message_type, sequence, monotonic_ns, sample_rate, channels, sample_width, plen = parsed
    if len(data) != HEADER_SIZE + plen:
        return None
    payload = data[HEADER_SIZE : HEADER_SIZE + plen]
    if message_type == CaptureMessageType.PCM and (
        not payload or len(payload) % (channels * sample_width) != 0
    ):
        return None
    if message_type in {CaptureMessageType.END, CaptureMessageType.CANCEL_ACK} and payload:
        return None
    if message_type == CaptureMessageType.ERROR and len(payload) > 64:
        return None
    if message_type == CaptureMessageType.READY and len(payload) > 1024:
        return None
    return DecodedCaptureFrame(
        message_type=message_type,
        sequence=sequence,
        monotonic_ns=monotonic_ns,
        sample_rate=sample_rate,
        channels=channels,
        sample_width=sample_width,
        payload=payload,
    )

File: native/macos_audio_capture/Sources/MacOSAudioCaptureLib/Framing.swift
import Foundation

/// Binary framing protocol between hikari-macos-audio-capture and Python.
/// Little-endian, fixed 48-byte header + bounded payload.
/// Layout:
/// magic(4) version(u16) type(u16) sequence(u64) monotonicNs(u64)
/// sampleRate(u32) channels(u16) sampleWidth(u16) payloadLength(u32)
/// reserved0(u32) reserved1(u32) reserved2(u32)  => 48 bytes
public enum CaptureFrameType: UInt16 {
    case ready = 1
    case pcm = 2
    case error = 3
    case end = 4
    case cancelAck = 5
}

public struct CaptureFrameHeader {
    public static let magic = Data([0x48, 0x49, 0x4B, 0x41]) // "HIKA"
    public static let version: UInt16 = 1
    public static let size = 48
    public static let maxPayload = 65_536

    public var messageType: CaptureFrameType
    public var sequence: UInt64
    public var monotonicNs: UInt64
    public var sampleRate: UInt32
    public var channels: UInt16
    public var sampleWidth: UInt16
    public var payloadLength: UInt32

    public init(
        messageType: CaptureFrameType,
        sequence: UInt64,
        monotonicNs: UInt64,
        sampleRate: UInt32 = 16_000,
        channels: UInt16 = 1,
        sampleWidth: UInt16 = 2,
        payloadLength: UInt32 = 0
    ) {
        self.messageType = messageType
        self.sequence = sequence
        self.monotonicNs = monotonicNs
        self.sampleRate = sampleRate
        self.channels = channels
        self.sampleWidth = sampleWidth
        self.payloadLength = payloadLength
    }

    public func encode() -> Data {
        var data = Data()
        data.reserveCapacity(Self.size)
        data.append(Self.magic)
        func appendU16(_ v: UInt16) {
            var le = v.littleEndian
            withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
        }
        func appendU32(_ v: UInt32) {
            var le = v.littleEndian
            withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
        }
        func appendU64(_ v: UInt64) {
            var le = v.littleEndian
            withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
        }
        appendU16(Self.version)
        appendU16(messageType.rawValue)
        appendU64(sequence)
        appendU64(monotonicNs)
        appendU32(sampleRate)
        appendU16(channels)
        appendU16(sampleWidth)
        appendU32(payloadLength)
        appendU32(0)
        appendU32(0)
        appendU32(0)
        precondition(data.count == Self.size)
        return data
    }

    public static func decode(_ data: Data) -> CaptureFrameHeader? {
        guard data.count >= size else { return nil }
        guard data.prefix(4) == magic else { return nil }
        func u16(_ o: Int) -> UInt16 {
            UInt16(data[o]) | (UInt16(data[o + 1]) << 8)
        }
        func u32(_ o: Int) -> UInt32 {
            UInt32(data[o])
                | (UInt32(data[o + 1]) << 8)
                | (UInt32(data[o + 2]) << 16)
                | (UInt32(data[o + 3]) << 24)
        }
        func u64(_ o: Int) -> UInt64 {
            var value: UInt64 = 0
            for index in 0..<8 {
                value |= UInt64(data[o + index]) << UInt64(index * 8)
            }
            return value
        }
        let version = u16(4)
        guard version == Self.version else { return nil }
        guard let type = CaptureFrameType(rawValue: u16(6)) else { return nil }
        let plen = u32(32)
        guard plen <= maxPayload else { return nil }
        guard u32(36) == 0, u32(40) == 0, u32(44) == 0 else { return nil }
        guard u32(24) == 16_000, u16(28) == 1, u16(30) == 2 else { return nil }
        return CaptureFrameHeader(
            messageType: type,
            sequence: u64(8),
            monotonicNs: u64(16),
            sampleRate: u32(24),
            channels: u16(28),
            sampleWidth: u16(30),
            payloadLength: plen
        )
    }
}

public enum CaptureFraming {
    public static func encodeFrame(header: CaptureFrameHeader, payload: Data) -> Data? {
        guard payload.count == Int(header.payloadLength),
              payload.count <= CaptureFrameHeader.maxPayload else { return nil }
        var out = header.encode()
        out.append(payload)
        return out
    }
}
